### PR TITLE
Require that LUBs and GLBs be unique

### PR DESCRIPTION
**Pull request by mernst**

### CODE REVIEW

**Comment by mernst** [comment]
This makes the Checker Framework will produce a less obscure error for checker implementation bugs like that in #2060.

**Comment by mernst** [comment]
The only significant change in this pull request is to issue an error when there is not a unique lub or glb, instead of returning an arbitrary value.  The pull request also eliminates duplication by factoring common logic into a method.

File: framework/src/main/java/org/checkerframework/framework/util/MultiGraphQualifierHierarchy.java [hunk 1]
```diff
@@ -926,11 +880,70 @@ private AnnotationMirror findGlbWithPoly(AnnotationMirror poly, AnnotationMirror
...
+            ErrorReporter.errorAbort(
+                    "MultiGraphQualifierHierarchy could not determine "
+                            + (lub ? "LUB" : "GLB")
+                            + " for "
+                            + a1
+                            + " and "
+                            + a2
+                            + ". Please ensure that the checker knows about all type qualifiers.");
+            return null;
+        } else if (outset.size() == 1) {
```
[the remark this answers is not in the document]
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/util/MultiGraphQualifierHierarchy.java]: Currently, it's in order, making it easy to see that the logic is exhaustive.
Furthermore, putting short cases earlier makes the flow easier to read.
I found the previous order more confusing and this is an intentional change.

File: framework/src/main/java/org/checkerframework/framework/util/MultiGraphQualifierHierarchy.java [hunk 2]
```diff
@@ -926,11 +880,70 @@ private AnnotationMirror findGlbWithPoly(AnnotationMirror poly, AnnotationMirror
...
+                if (isPolymorphicQualifier(anno)) {
+                    continue;
+                } else if (result == null) {
+                    result = anno;
+                } else {
+                    ErrorReporter.errorAbort(
+                            String.format(
+                                    "Bug in checker implementation:  type hierarchy is not a lattice.%n"
+                                            + "There is no unique "
+                                            + (lub ? "lub" : "glb")
```
[the remark this answers is not in the document]
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/util/MultiGraphQualifierHierarchy.java]: One is an English abbreviation.  The other is a mathematical function.  Different capitalization rules apply.

File: framework/src/main/java/org/checkerframework/framework/util/MultiGraphQualifierHierarchy.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/util/MultiGraphQualifierHierarchy.java]: The case for size 1 is the non-error case and is just one line long.
Handling the non-error cases first is easier to read.
I'm confused about the "putting short cases earlier" argument - the case for size 0 is 10 lines and now before the case for size 1 which is 1 line.

File: framework/src/main/java/org/checkerframework/framework/util/MultiGraphQualifierHierarchy.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/util/MultiGraphQualifierHierarchy.java]: The argument was that the two error messages use different formats, one an English abbreviation the other a mathematical function.
Why not pick one and use that consistently in both cases?

[the remark this answers is not in the document]
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/util/MultiGraphQualifierHierarchy.java]: I have no objection to doing so.

File: framework/src/main/java/org/checkerframework/framework/util/MultiGraphQualifierHierarchy.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/util/MultiGraphQualifierHierarchy.java]: Size > 1 is also a non-error case.

**Pull request merged**